### DISCUSSION TITLE
Help needed in linking the taxonomy with the Facts.

### DISCUSSION THREAD

**Discussion by galigutta** [manusimidt/py-xbrl#122] [Q&A]
Hi, I have looked around a bit, and I am struggling to figure out how to embed the facts in the calculation_link to get a view of the various tables in the filing. 

Specifically the problem I am running into is that the concepts of facts need to be combined with concept of the taxonomy.linked_roles.calculation_link under different contexts. How is this done? it does seem to get more complicated with near similar facts in the instance. Would anyone be kind enough to post their code to accomplish this?

**Comment by rayniervanegmond** [comment]
Ill do a writeup later today. This is something that XBRL has support for
but…. I need to catch a ferry to to to class … so TTFN

René (he/him)

**Reply by galigutta** [reply]
Thank you so much. Glad to know there is support to do this. Seems painful to handcraft myself.

**Reply by galigutta** [reply]
@rayniervanegmond  I was wondering if your comment was in reference to XBRL in general or this library in particular or both! I would ideally like to use it in the context of this library to be sure.

**Reply by elliotdes** [reply]
@rayniervanegmond This is something I would also be interested in!

**Comment by galigutta** [comment]
I think I solved this. here for posterity.

Of course if any one know a better way please chime in. This was more straightforward than I was thinking.

```python
def get_facts_by_xml_id(inst, xml_id):
    return [fact for fact in inst.facts if fact.concept.xml_id == xml_id]

def get_label_by_xml_id(inst, xml_id):
    for i in inst.taxonomy.lab_linkbases[0].extended_links[0].root_locators:
        if i.concept_id == xml_id:
            return(i.children[0].labels[0].text)
    
# concept_str comes from linkbases
concept_str = 'us-gaap_RevenueFromContractWithCustomerExcludingAssessedTax'
filtered_facts = get_facts_by_xml_id(inst, concept_str)
label = get_label_by_xml_id(inst, concept_str)

print(label)
for fact in filtered_facts:
    if len(fact.context.segments) == 0:
        print(fact.value)
        print(fact.context)
```
